crates/types/src/replication/nodeset_selector.rs

```diff
@@ -0,0 +1,1052 @@
+        for i in 1..=100 {
+            nodes_config.upsert_node(generate_node(i, StorageState::Disabled, Role::Worker, ""));
+        }
+
+        // replication = 2, we'll try an have at least 2 writeable nodes in the nodeset. So, we
+        // should expect 5, 8 both to be here.
+        let replication =
+            ReplicationProperty::with_scope(LocationScope::Node, 2.try_into().unwrap());
+        // simulates that we consider 5, 8 are writeable nodes via cluster_state, 5,6 are alive, we prefer to see them in nodeset
+        let cluster_state = NodeSet::from_iter([5, 8]);
+
+        let preferred_nodes = NodeSet::from_single(1);
+        let options = NodeSetSelectorOptions::new(5).with_preferred_nodes(&preferred_nodes);
+        let nodeset = DomainAwareNodeSetSelector::select(
+            &nodes_config,
+            &replication,
+            worker_candidate_filter,
+            |node_id, _| cluster_state.contains(node_id),
+            options.clone(),
+        )
+        .unwrap();
+
+        assert!(nodeset.contains(5));
+        assert!(nodeset.contains(8));
+        assert!(nodeset.contains(1));
+        assert_eq!(nodeset.len(), 5);
+
+        // We can also generate bigger nodesets other than the default size
+        let options = options.with_target_size(10);
+        let nodeset = DomainAwareNodeSetSelector::select(
+            &nodes_config,
+            &replication,
+            worker_candidate_filter,
+            |node_id, _| cluster_state.contains(node_id),
+            options,
+        )
+        .unwrap();
+
+        assert!(nodeset.contains(5));
+        assert!(nodeset.contains(8));
+        assert!(nodeset.contains(1));
+        // target size is not a strict number, but we try to stay near it.
+        assert_eq!(nodeset.len(), 12);
+    }
+
+    #[test]
+    fn select_fd_select_nodes() {
+        let mut nodes_config = NodesConfiguration::default();
+        for i in 1..=12 {
+            // three regions, 2 zones per region
+            let region = i % 3;
+            let zone = i % 2;
+            nodes_config.upsert_node(generate_node(
+                i,
+                StorageState::ReadWrite,
+                Role::Worker,
+                &format!("region-{region}.zone-{zone}"),
+            ));
+        }
+
+        // zone:2. we have 3 * 2 = 6 zones in total.
+        let replication: ReplicationProperty = "{zone: 2}".parse().unwrap();
+
+        // we like node 1
+        let preferred_nodes = NodeSet::from_single(1);
+        let options = NodeSetSelectorOptions::new(5).with_preferred_nodes(&preferred_nodes);
+        let nodeset = DomainAwareNodeSetSelector::select(
+            &nodes_config,
+            &replication,
+            logserver_candidate_filter,
+            logserver_writeable_node_filter,
+            options,
+        )
+        .unwrap();
+
+        assert!(nodeset.contains(1));
+        assert_eq!(nodeset.len(), 6);
```

Comment:
The nodeset size is 6 because we run the `choose_domains` twice (since all nodes are effectively writable), right?

Reply:
It's 6 because we try and get nodes from every domain possible to increase write availability.